### PR TITLE
Add Diff::equals

### PR DESCRIPTION
**Pull request by thiemowmde**
Note that I'm not using the `Comparable` interface here to not introduce a weird dependency that's not crucial.

I squished some basic refactoring in here. I can split this into separate patches if you like. No problem, just tell me.
* Inline two trivial methods (looks like they were overriding `protected` parent methods before but don't do any more).
* Some more specific type hints.
* Favor `bool` and `int` shortcuts.

Bug: [T92487](https://phabricator.wikimedia.org/T92487).

### CODE REVIEW

File: tests/phpunit/DiffOp/Diff/DiffTest.php [hunk 1]
```diff
@@ -673,4 +672,54 @@ public function invalidDiffOpsProvider() {
...
+	public function notEqualsProvider() {
+		return array(
+			// Empty diffs
+			array( new Diff(), new Diff( array(), false ) ),
+			array( new Diff(), new Diff( array(), true ) ),
+
+			// Simple diffs
+			array( new Diff(), new Diff( array( new DiffOpAdd( 1 ) ) ) ),
+			array( new Diff( array( new DiffOpAdd( 1 ) ) ), new Diff( array( new DiffOpRemove( 1 ) ) ) ),
+			array( new Diff( array( new DiffOpAdd( 1 ) ) ), new Diff( array( new DiffOpAdd( 2 ) ) ) ),
```
Comment (thiemowmde) [inline tests/phpunit/DiffOp/Diff/DiffTest.php]: This single line is the relevant test for [T92487](https://phabricator.wikimedia.org/T92487).

File: tests/phpunit/DiffOp/Diff/DiffTest.php [hunk 2]
```diff
@@ -673,4 +672,54 @@ public function invalidDiffOpsProvider() {
...
+
+			// Subclasses
+			// TODO: Add a test with an EntityContentDiff?
+			// TODO: Add a test with an EntityDiff?
+			// TODO: Add a test with an ItemDiff?
```
Comment (JeroenDeDauw) [inline tests/phpunit/DiffOp/Diff/DiffTest.php]: Please remove these comments, they refer to constructs not part of this component or one of its dependencies

File: src/DiffOp/Diff/Diff.php [hunk 3]
```diff
@@ -71,15 +73,6 @@ public function __construct( array $operations = array(), $isAssociative = null
 	}
 
 	/**
-	 * Returns the name of an interface/class that the element should implement/extend.
-	 *
-	 * @return string
-	 */
-	private function getObjectType() {
-		return '\Diff\DiffOp\DiffOp';
-	}
```
Comment (JeroenDeDauw) [inline src/DiffOp/Diff/Diff.php]: Hah. Surprised this was still here. Been a long time since this class stopped deriving from GenericArrayObject.

**Comment by JeroenDeDauw** [comment]
Code looks good. Don't forget the release notes.

File: src/DiffOp/Diff/Diff.php [hunk 4]
```diff
@@ -366,11 +356,31 @@ public function toArray( $valueConverter = null ) {
 	}
 
 	/**
+	 * @since 2.0
+	 *
+	 * @param mixed $target
+	 *
+	 * @return bool
+	 */
+	public function equals( $target ) {
```
Comment (JeroenDeDauw) [inline src/DiffOp/Diff/Diff.php]: You could also implement this by doing a diff between the two diffs and seeing if it is equal to an empty diff.

inception

Comment (thiemowmde) [inline src/DiffOp/Diff/Diff.php]: I'm afraid this is what PHP does when [comparing two ArrayObjects](https://phabricator.wikimedia.org/T92487#1113096).

File: tests/phpunit/DiffOp/Diff/DiffTest.php (hunk 2 above)
Comment (thiemowmde) [inline tests/phpunit/DiffOp/Diff/DiffTest.php]: Oh yes, this was stupid, I should have checked where these classes live. Will remove.

**Pull request merged**